README.md
# flat-tree

[![crates.io version][1]][2] [![build status][3]][4]
[![downloads][5]][6] [![docs.rs docs][7]][8]

Map a binary tree to a list. Adapted from
[mafintosh/flat-tree](https://github.com/mafintosh/flat-tree).

- [Documentation][8]
- [Crates.io][2]

## Usage
```rust
extern crate flat_tree;

let parent = flat_tree::parent(0);
println!("parent of 0 is {}", parent);
```

## Why?
You can represent a binary tree in a simple flat list using the following
structure:

```text
      3
  1       5
0   2   4   6  ...
```

Each number represents an **index** in a flat list. So a tree:

```text
      A
  B       C
D   E   F   G  ...
```

would be represented as a list: `[D B E A F C G]`

Furthermore, indexes `0`, `2`, `4`, `6` are on **depth** `0`. `1`, `5`, `9` on depth `1`. And so forth.

```text
depth = 2  ^        3
depth = 1  |    1       5
depth = 0  |  0   2   4   6  ...
```

In some cases it is also useful to calculate an **offset**. Indexes `0`, `1`, `3`, `7` have an offset `0`:

```text
                (7)
       (3)
  (1)       5
(0)   2   4   6      ...
```

`2`, `5`, `11`, `23` offset `1`:

```text
                 7
       3                  (11)
  1        (5)        9          13
0   (2)   4   6    10   12    14    15
```

This module exposes a series of functions to help you build and maintain
this data structure.

## License
[MIT](./LICENSE-MIT) OR [Apache-2.0](./LICENSE-APACHE)

[1]: https://img.shields.io/crates/v/flat-tree.svg?style=flat-square
[2]: https://crates.io/crates/flat-tree
[3]: https://img.shields.io/travis/datrs/flat-tree.svg?style=flat-square
[4]: https://travis-ci.org/datrs/flat-tree
[5]: https://img.shields.io/crates/d/flat-tree.svg?style=flat-square
[6]: https://crates.io/crates/flat-tree
[7]: https://docs.rs/flat-tree/badge.svg
[8]: https://docs.rs/flat-tree


CHANGELOG.md
## 2020-03-03, Version 5.0.0
### Commits
- [[`05ffe8699d`](https://github.com/mafintosh/flat-tree-rs/commit/05ffe8699d452b01152a596e5cb0a652e7915503)] (cargo-release) version 5.0.0 (Bruno Tavares)
- [[`709b2d5635`](https://github.com/mafintosh/flat-tree-rs/commit/709b2d5635c4d2f9fdad48e23d55d257294dbdd9)] Merge pull request #44 from bltavares/usize-to-u64 (Bruno Tavares)
- [[`bc9c866bdb`](https://github.com/mafintosh/flat-tree-rs/commit/bc9c866bdb2101aa892b4bf917a8b1a30be1da45)] Fix travis script (Bruno Tavares)
- [[`68f7e67ed4`](https://github.com/mafintosh/flat-tree-rs/commit/68f7e67ed4c76fc6c11bb45ea78e9ec41b3d2449)] Change from usize to u64 (Bruno Tavares)
- [[`f9ca2ef68e`](https://github.com/mafintosh/flat-tree-rs/commit/f9ca2ef68e19a861117788bb6624e9529b1d6419)] Update criterion requirement from 0.2 to 0.3 (dependabot-preview[bot])
- [[`b23328579d`](https://github.com/mafintosh/flat-tree-rs/commit/b23328579d00947d6480c3e894e18e1598905204)] Update changelog (Yoshua Wuyts)

### Stats
```diff
 .travis.yml     |  6 +++---
 CHANGELOG.md    | 15 +++++++++++++++
 Cargo.toml      |  4 ++--
 src/iterator.rs | 36 ++++++++++++++++++------------------
 src/lib.rs      | 34 +++++++++++++++++-----------------
 5 files changed, 55 insertions(+), 40 deletions(-)
```


## 2016-03-09, Version v1.0.0
### Commits
- [[`b256517aef`](https://github.com/mafintosh/flat-tree-rs/commit/b256517aefb4ce2092e2920931d4c78cca6590af)] add badge (Mathias Buus)
- [[`017436cbae`](https://github.com/mafintosh/flat-tree-rs/commit/017436cbaec00ef57b7e8311a5f674fd0cf418e4)] add desc (Mathias Buus)
- [[`388d2445bc`](https://github.com/mafintosh/flat-tree-rs/commit/388d2445bc567148867a79f315aeee3ca492b100)] first commit (Mathias Buus)

### Stats
```diff
 Cargo.toml   |   2 +-
 rustfmt.toml |   2 +-
 src/lib.rs   | 328 ++++++++++++++++++++++++++----------------------------------
 3 files changed, 144 insertions(+), 188 deletions(-)
```


## 2016-03-09, Version v1.0.0
### Commits
- [[`b256517aef`](https://github.com/mafintosh/flat-tree-rs/commit/b256517aefb4ce2092e2920931d4c78cca6590af)] add badge (Mathias Buus)
- [[`017436cbae`](https://github.com/mafintosh/flat-tree-rs/commit/017436cbaec00ef57b7e8311a5f674fd0cf418e4)] add desc (Mathias Buus)
- [[`388d2445bc`](https://github.com/mafintosh/flat-tree-rs/commit/388d2445bc567148867a79f315aeee3ca492b100)] first commit (Mathias Buus)

### Stats
```diff
 Cargo.toml   |   2 +-
 rustfmt.toml |   2 +-
 src/lib.rs   | 328 ++++++++++++++++++++++++++----------------------------------
 3 files changed, 144 insertions(+), 188 deletions(-)
```


## 2016-03-09, Version v1.0.0
### Commits
- [[`b256517aef`](https://github.com/mafintosh/flat-tree-rs/commit/b256517aefb4ce2092e2920931d4c78cca6590af)] add badge (Mathias Buus)
- [[`017436cbae`](https://github.com/mafintosh/flat-tree-rs/commit/017436cbaec00ef57b7e8311a5f674fd0cf418e4)] add desc (Mathias Buus)
- [[`388d2445bc`](https://github.com/mafintosh/flat-tree-rs/commit/388d2445bc567148867a79f315aeee3ca492b100)] first commit (Mathias Buus)

### Stats
```diff
 Cargo.toml   |   2 +-
 rustfmt.toml |   2 +-
 src/lib.rs   | 328 ++++++++++++++++++++++++++----------------------------------
 3 files changed, 144 insertions(+), 188 deletions(-)
```


## 2016-03-09, Version v1.0.0
### Commits
- [[`b256517aef`](https://github.com/mafintosh/flat-tree-rs/commit/b256517aefb4ce2092e2920931d4c78cca6590af)] add badge (Mathias Buus)
- [[`017436cbae`](https://github.com/mafintosh/flat-tree-rs/commit/017436cbaec00ef57b7e8311a5f674fd0cf418e4)] add desc (Mathias Buus)
- [[`388d2445bc`](https://github.com/mafintosh/flat-tree-rs/commit/388d2445bc567148867a79f315aeee3ca492b100)] first commit (Mathias Buus)

### Stats
```diff
 Cargo.toml   |   2 +-
 rustfmt.toml |   2 +-
 src/lib.rs   | 328 ++++++++++++++++++++++++++----------------------------------
 3 files changed, 144 insertions(+), 188 deletions(-)
```


## 2016-03-09, Version v1.0.0
### Commits
- [[`b256517aef`](https://github.com/mafintosh/flat-tree-rs/commit/b256517aefb4ce2092e2920931d4c78cca6590af)] add badge (Mathias Buus)
- [[`017436cbae`](https://github.com/mafintosh/flat-tree-rs/commit/017436cbaec00ef57b7e8311a5f674fd0cf418e4)] add desc (Mathias Buus)
- [[`388d2445bc`](https://github.com/mafintosh/flat-tree-rs/commit/388d2445bc567148867a79f315aeee3ca492b100)] first commit (Mathias Buus)

### Stats
```diff
 Cargo.toml   |   2 +-
 rustfmt.toml |   2 +-
 src/lib.rs   | 328 ++++++++++++++++++++++++++----------------------------------
 3 files changed, 144 insertions(+), 188 deletions(-)
```


## 2016-03-09, Version v1.0.0
### Commits
- [[`b256517aef`](https://github.com/mafintosh/flat-tree-rs/commit/b256517aefb4ce2092e2920931d4c78cca6590af)] add badge (Mathias Buus)
- [[`017436cbae`](https://github.com/mafintosh/flat-tree-rs/commit/017436cbaec00ef57b7e8311a5f674fd0cf418e4)] add desc (Mathias Buus)
- [[`388d2445bc`](https://github.com/mafintosh/flat-tree-rs/commit/388d2445bc567148867a79f315aeee3ca492b100)] first commit (Mathias Buus)

### Stats
```diff
 Cargo.toml   |   2 +-
 rustfmt.toml |   2 +-
 src/lib.rs   | 328 ++++++++++++++++++++++++++----------------------------------
 3 files changed, 144 insertions(+), 188 deletions(-)
```




tests/iterator.rs
extern crate flat_tree;

#[test]
fn iterator() {
  let mut iterator = flat_tree::Iterator::default();
  assert_eq!(iterator.index(), 0);
  assert_eq!(iterator.prev(), 0);
  assert_eq!(iterator.is_left(), true);
  assert_eq!(iterator.is_right(), false);
  assert_eq!(iterator.parent(), 1);
  assert_eq!(iterator.offset(), 0);
  assert_eq!(iterator.parent(), 3);
  assert_eq!(iterator.offset(), 0);
  assert_eq!(iterator.parent(), 7);
  assert_eq!(iterator.offset(), 0);
  assert_eq!(iterator.right_child(), 11);
  assert_eq!(iterator.left_child(), 9);
  assert_eq!(iterator.next(), Some(13));
  assert_eq!(iterator.is_left(), false);
  assert_eq!(iterator.is_right(), true);
  assert_eq!(iterator.left_span(), 12);
  assert_eq!(iterator.next(), Some(14));
  assert_eq!(iterator.next(), Some(16));
  assert_eq!(iterator.offset(), 8);
  assert_eq!(iterator.parent(), 17);
  assert_eq!(iterator.offset(), 4);
  assert_eq!(iterator.parent(), 19);
  assert_eq!(iterator.offset(), 2);
  assert_eq!(iterator.parent(), 23);
  assert_eq!(iterator.offset(), 1);
  assert_eq!(iterator.right_span(), 30);

  iterator.seek(23);

  assert_eq!(iterator.index(), 23);
  assert_eq!(iterator.offset(), 1);
  assert_eq!(iterator.right_child(), 27);
  assert_eq!(iterator.sibling(), 19);
  assert_eq!(iterator.prev(), 11);
  assert_eq!(iterator.left_child(), 9);
  assert_eq!(iterator.prev(), 5);
  assert_eq!(iterator.left_child(), 4);
  assert_eq!(iterator.prev(), 2);
  assert_eq!(iterator.prev(), 0);
}

#[test]
fn non_leaf_start() {
  let mut iterator = flat_tree::Iterator::new(1);
  assert_eq!(iterator.index(), 1);
  assert_eq!(iterator.parent(), 3);
  assert_eq!(iterator.parent(), 7);
  assert_eq!(iterator.right_child(), 11);
  assert_eq!(iterator.left_child(), 9);
  assert_eq!(iterator.next(), Some(13));
  assert_eq!(iterator.left_span(), 12);
}


tests/iterator_regression.rs
extern crate flat_tree;

#[test]
/// Postmortem: offset was incorrectly calculated when finding a parent for a
/// node with an odd offset.
fn parent_and_odd_offset() {
  let mut iterator = flat_tree::Iterator::new(10);
  assert_eq!(iterator.index(), 10);
  assert_eq!(iterator.offset(), 5);
  assert_eq!(iterator.parent(), 9);
  assert_eq!(iterator.offset(), 2);
  assert_eq!(iterator.parent(), 11);
  assert_eq!(iterator.offset(), 1);
}


src/iterator.rs
//! ## Usage
//! ```rust
//! let mut iter = flat_tree::Iterator::new(0);
//! assert_eq!(iter.next(), Some(2));
//! assert_eq!(iter.next(), Some(4));
//! assert_eq!(iter.next(), Some(6));
//! assert_eq!(iter.parent(), 5);
//! ```
use super::*;

use std::iter;

/// Iterator over a flat-tree.
#[derive(Debug)]
pub struct Iterator {
  index: u64,
  offset: u64,
  factor: u64,
}

impl Iterator {
  /// Create a new iterator.
  ///
  /// ## Examples
  /// ```rust
  /// use flat_tree::Iterator;
  /// assert_eq!(Iterator::new(0).take(3).collect::<Vec<u64>>(), [2, 4, 6]);
  /// ```
  #[inline]
  pub fn new(index: u64) -> Self {
    let mut instance = Self {
      index: 0,
      offset: 0,
      factor: 0,
    };

    instance.seek(index);
    instance
  }

  /// Get the current index.
  #[inline]
  pub fn index(&self) -> u64 {
    self.index
  }

  /// Get the current offset.
  #[inline]
  pub fn offset(&self) -> u64 {
    self.offset
  }

  /// Get the current factor.
  #[inline]
  pub fn factor(&self) -> u64 {
    self.factor
  }

  /// Seek to a position in the iterator.
  ///
  /// ## Examples
  /// ```rust
  /// let mut iter = flat_tree::Iterator::new(0);
  /// iter.seek(4);
  /// assert_eq!(iter.next(), Some(6));
  /// iter.seek(2);
  /// assert_eq!(iter.next(), Some(4));
  /// ```
  #[inline]
  pub fn seek(&mut self, index: u64) {
    self.index = index;
    if is_odd(self.index) {
      self.offset = offset(index);
      self.factor = two_pow(depth(index) + 1);
    } else {
      self.offset = index / 2;
      self.factor = 2;
    }
  }

  /// Check if the position of the iterator is currently on a left node.
  ///
  /// ## Examples
  /// ```rust
  /// assert_eq!(flat_tree::Iterator::new(0).is_left(), true);
  /// assert_eq!(flat_tree::Iterator::new(2).is_left(), false);
  /// assert_eq!(flat_tree::Iterator::new(1).is_left(), true);
  /// ```
  #[inline]
  pub fn is_left(&self) -> bool {
    is_even(self.offset)
  }

  /// Check if the position of the iterator is currently on a right node.
  ///
  /// ## Examples
  /// ```rust
  /// assert_eq!(flat_tree::Iterator::new(0).is_right(), false);
  /// assert_eq!(flat_tree::Iterator::new(2).is_right(), true);
  /// assert_eq!(flat_tree::Iterator::new(1).is_right(), false);
  /// ```
  #[inline]
  pub fn is_right(&self) -> bool {
    is_odd(self.offset)
  }

  /// Move the cursor and get the previous item from the current position.
  ///
  /// ## Examples
  /// ```rust
  /// let mut iter = flat_tree::Iterator::new(6);
  /// assert_eq!(iter.prev(), 4);
  /// assert_eq!(iter.prev(), 2);
  /// assert_eq!(iter.prev(), 0);
  /// ```
  #[inline]
  pub fn prev(&mut self) -> u64 {
    if self.offset == 0 {
      return self.index;
    }
    self.offset -= 1;
    self.index -= self.factor;
    self.index
  }

  /// Get the sibling for the current position and move the cursor.
  ///
  /// ## Examples
  /// ```rust
  /// assert_eq!(flat_tree::Iterator::new(0).sibling(), 2);
  /// assert_eq!(flat_tree::Iterator::new(1).sibling(), 5);
  /// assert_eq!(flat_tree::Iterator::new(4).sibling(), 6);
  /// ```
  #[inline]
  pub fn sibling(&mut self) -> u64 {
    if self.is_left() {
      self.next().unwrap() // this is always safe
    } else {
      self.prev()
    }
  }

  /// Get the parent for the current position and move the cursor.
  ///
  /// ## Examples
  /// ```rust
  /// assert_eq!(flat_tree::Iterator::new(0).parent(), 1);
  /// assert_eq!(flat_tree::Iterator::new(1).parent(), 3);
  /// assert_eq!(flat_tree::Iterator::new(4).parent(), 5);
  /// ```
  #[inline]
  pub fn parent(&mut self) -> u64 {
    if is_odd(self.offset) {
      self.index -= self.factor / 2;
      self.offset = (self.offset - 1) / 2;
    } else {
      self.index += self.factor / 2;
      self.offset /= 2;
    }
    self.factor *= 2;
    self.index
  }

  /// Get the left_span for the current position and move the cursor.
  ///
  /// ## Examples
  /// ```rust
  /// assert_eq!(flat_tree::Iterator::new(0).left_span(), 0);
  /// assert_eq!(flat_tree::Iterator::new(1).left_span(), 0);
  /// assert_eq!(flat_tree::Iterator::new(3).left_span(), 0);
  /// assert_eq!(flat_tree::Iterator::new(23).left_span(), 16);
  /// assert_eq!(flat_tree::Iterator::new(27).left_span(), 24);
  /// ```
  #[inline]
  pub fn left_span(&mut self) -> u64 {
    self.index = self.index + 1 - self.factor / 2;
    self.offset = self.index / 2;
    self.factor = 2;
    self.index
  }

  /// Get the right_span for the current position and move the cursor.
  ///
  /// ## Examples
  /// ```rust
  /// assert_eq!(flat_tree::Iterator::new(0).right_span(), 0);
  /// assert_eq!(flat_tree::Iterator::new(1).right_span(), 2);
  /// assert_eq!(flat_tree::Iterator::new(3).right_span(), 6);
  /// assert_eq!(flat_tree::Iterator::new(23).right_span(), 30);
  /// assert_eq!(flat_tree::Iterator::new(27).right_span(), 30);
  /// ```
  #[inline]
  pub fn right_span(&mut self) -> u64 {
    self.index = self.index + self.factor / 2 - 1;
    self.offset = self.index / 2;
    self.factor = 2;
    self.index
  }

  /// Get the left_child for the current position and move the cursor.
  ///
  /// ## Examples
  /// ```rust
  /// assert_eq!(flat_tree::Iterator::new(1).left_child(), 0);
  /// assert_eq!(flat_tree::Iterator::new(3).left_child(), 1);
  /// assert_eq!(flat_tree::Iterator::new(7).left_child(), 3);
  /// ```
  #[inline]
  pub fn left_child(&mut self) -> u64 {
    if self.factor == 2 {
      return self.index;
    }
    self.factor /= 2;
    self.index -= self.factor / 2;
    self.offset *= 2;
    self.index
  }

  /// Get the right_child for the current position and move the cursor.
  ///
  /// ## Examples
  /// ```rust
  /// assert_eq!(flat_tree::Iterator::new(1).right_child(), 2);
  /// assert_eq!(flat_tree::Iterator::new(3).right_child(), 5);
  /// assert_eq!(flat_tree::Iterator::new(7).right_child(), 11);
  /// ```
  #[inline]
  pub fn right_child(&mut self) -> u64 {
    if self.factor == 2 {
      return self.index;
    }
    self.factor /= 2;
    self.index += self.factor / 2;
    self.offset = 2 * self.offset + 1;
    self.index
  }
}

impl iter::Iterator for Iterator {
  type Item = u64;

  #[inline]
  fn next(&mut self) -> Option<Self::Item> {
    self.offset += 1;
    self.index += self.factor;
    Some(self.index)
  }
}

impl Default for Iterator {
  #[inline]
  fn default() -> Self {
    Self::new(0)
  }
}

#[inline]
fn two_pow(n: u64) -> u64 {
  if n < 31 {
    1 << n
  } else {
    ((1 << 30) * (1 << (n - 30)))
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn test_two_pow() {
    assert_eq!(two_pow(0), 1);
    assert_eq!(two_pow(1), 2);
    assert_eq!(two_pow(2), 4);
    assert_eq!(two_pow(3), 8);
    assert_eq!(two_pow(31), 2147483648);
  }

  #[cfg(target_pointer_width = "64")]
  #[test]
  fn test_two_pow_64bit() {
    assert_eq!(two_pow(34), 17179869184);
    assert_eq!(two_pow(63), 9223372036854775808);
  }
}


benches/bench.rs
#[macro_use]
extern crate criterion;
extern crate flat_tree;

use criterion::Criterion;

fn criterion_benchmark(c: &mut Criterion) {
  c.bench_function("children_9", |b| b.iter(|| flat_tree::children(9)));
  c.bench_function("children_256", |b| b.iter(|| flat_tree::children(256)));
  c.bench_function("count_128", |b| b.iter(|| flat_tree::count(128)));
  c.bench_function("depth_128", |b| b.iter(|| flat_tree::depth(128)));
  // c.bench_function("full_roots_16", |b| {
  //   let mut nodes: Vec<usize> = Vec::with_capacity(16);
  //   b.iter(|| {
  //     flat_tree::full_roots(16, &mut nodes);
  //   })
  // });
  // c.bench_function("full_roots_256", |b| {
  //   let mut nodes: Vec<usize> = Vec::with_capacity(16);
  //   b.iter(|| {
  //     flat_tree::full_roots(256, &mut nodes);
  //   })
  // });
  c.bench_function("index_3_1", |b| b.iter(|| flat_tree::index(3, 1)));
  c.bench_function("left_child_128", |b| b.iter(|| flat_tree::left_child(128)));
  c.bench_function("left_span_128", |b| b.iter(|| flat_tree::left_span(128)));
  c.bench_function("offset_128", |b| b.iter(|| flat_tree::offset(128)));
  c.bench_function("parent_128", |b| b.iter(|| flat_tree::parent(128)));
  c.bench_function("right_child_128", |b| {
    b.iter(|| flat_tree::right_child(128))
  });
  c.bench_function("right_span_128", |b| b.iter(|| flat_tree::right_span(128)));
  c.bench_function("sibling_128", |b| b.iter(|| flat_tree::sibling(128)));
  c.bench_function("spans_128", |b| b.iter(|| flat_tree::spans(128)));
  c.bench_function("uncle_128", |b| b.iter(|| flat_tree::uncle(128)));
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);


src/lib.rs
#![cfg_attr(feature = "nightly", deny(missing_docs))]
#![cfg_attr(feature = "nightly", feature(external_doc))]
#![cfg_attr(feature = "nightly", doc(include = "../README.md"))]

mod iterator;

pub use iterator::Iterator;

/// Returns the flat-tree of the tree node at the specified depth and offset.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::index(0, 0), 0);
/// assert_eq!(flat_tree::index(0, 1), 2);
/// assert_eq!(flat_tree::index(0, 2), 4);
/// assert_eq!(flat_tree::index(1, 2), 9);
/// assert_eq!(flat_tree::index(1, 3), 13);
/// assert_eq!(flat_tree::index(2, 1), 11);
/// assert_eq!(flat_tree::index(2, 2), 19);
/// assert_eq!(flat_tree::index(3, 0), 7);
/// assert_eq!(flat_tree::index(3, 1), 23);
/// ```
#[inline]
pub const fn index(depth: u64, offset: u64) -> u64 {
  (offset << (depth + 1)) | ((1 << depth) - 1)
}

/// Returns the depth of a node.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::depth(0), 0);
/// assert_eq!(flat_tree::depth(1), 1);
/// assert_eq!(flat_tree::depth(2), 0);
/// assert_eq!(flat_tree::depth(3), 2);
/// assert_eq!(flat_tree::depth(4), 0);
/// ```
#[inline]
pub const fn depth(i: u64) -> u64 {
  // Count trailing `1`s of the binary representation of the number.
  (!i).trailing_zeros() as u64
}

/// Returns the offset of a node.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::offset(0), 0);
/// assert_eq!(flat_tree::offset(1), 0);
/// assert_eq!(flat_tree::offset(2), 1);
/// assert_eq!(flat_tree::offset(3), 0);
/// assert_eq!(flat_tree::offset(4), 2);
/// ```
#[inline]
pub fn offset(i: u64) -> u64 {
  let depth = self::depth(i);
  if is_even(i) {
    i / 2
  } else {
    i >> (depth + 1)
  }
}

/// Returns the parent of a node with a depth.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::parent(0), 1);
/// assert_eq!(flat_tree::parent(1), 3);
/// assert_eq!(flat_tree::parent(2), 1);
/// assert_eq!(flat_tree::parent(3), 7);
/// assert_eq!(flat_tree::parent(4), 5);
/// ```
#[inline]
pub fn parent(i: u64) -> u64 {
  let depth = self::depth(i);
  index(depth + 1, offset(i) >> 1)
}

/// Returns the sibling of a node.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::sibling(0), 2);
/// assert_eq!(flat_tree::sibling(2), 0);
/// assert_eq!(flat_tree::sibling(1), 5);
/// assert_eq!(flat_tree::sibling(5), 1);
/// ```
#[inline]
pub fn sibling(i: u64) -> u64 {
  let depth = self::depth(i);
  index(depth, offset(i) ^ 1)
}

/// Returns the parent's sibling, of a node.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::uncle(0), 5);
/// assert_eq!(flat_tree::uncle(2), 5);
/// assert_eq!(flat_tree::uncle(1), 11);
/// assert_eq!(flat_tree::uncle(5), 11);
/// ```
#[inline]
pub fn uncle(i: u64) -> u64 {
  let depth = self::depth(i);
  index(depth + 1, offset(parent(i)) ^ 1)
}

/// Returns both children of a node.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::children(0), None);
/// assert_eq!(flat_tree::children(1), Some((0, 2)));
/// assert_eq!(flat_tree::children(3), Some((1, 5)));
/// assert_eq!(flat_tree::children(9), Some((8, 10)));
/// ```
#[inline]
pub fn children(i: u64) -> Option<(u64, u64)> {
  let depth = self::depth(i);
  if is_even(i) {
    None
  } else if depth == 0 {
    Some((i, i))
  } else {
    let offset = offset(i) * 2;
    Some((index(depth - 1, offset), index(depth - 1, offset + 1)))
  }
}

/// Returns only the left child of a node.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::left_child(0), None);
/// assert_eq!(flat_tree::left_child(1), Some(0));
/// assert_eq!(flat_tree::left_child(3), Some(1));
/// ```
#[inline]
pub fn left_child(i: u64) -> Option<u64> {
  let depth = self::depth(i);
  if is_even(i) {
    None
  } else if depth == 0 {
    Some(i)
  } else {
    Some(index(depth - 1, offset(i) << 1))
  }
}

/// Returns only the left child of a node.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::right_child(0), None);
/// assert_eq!(flat_tree::right_child(1), Some(2));
/// assert_eq!(flat_tree::right_child(3), Some(5));
/// ```
// TODO: handle errors
#[inline]
pub fn right_child(i: u64) -> Option<u64> {
  let depth = self::depth(i);
  if is_even(i) {
    None
  } else if depth == 0 {
    Some(i)
  } else {
    Some(index(depth - 1, (offset(i) << 1) + 1))
  }
}

/// Returns the right most node in the tree that the node spans.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::right_span(0), 0);
/// assert_eq!(flat_tree::right_span(1), 2);
/// assert_eq!(flat_tree::right_span(3), 6);
/// assert_eq!(flat_tree::right_span(23), 30);
/// assert_eq!(flat_tree::right_span(27), 30);
/// ```
#[inline]
pub fn right_span(i: u64) -> u64 {
  let depth = self::depth(i);
  if depth == 0 {
    i
  } else {
    (offset(i) + 1) * (2 << depth) - 2
  }
}

/// Returns the left most node in the tree that it spans.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::left_span(0), 0);
/// assert_eq!(flat_tree::left_span(1), 0);
/// assert_eq!(flat_tree::left_span(3), 0);
/// assert_eq!(flat_tree::left_span(23), 16);
/// assert_eq!(flat_tree::left_span(27), 24);
/// ```
#[inline]
pub fn left_span(i: u64) -> u64 {
  let depth = self::depth(i);
  if depth == 0 {
    i
  } else {
    offset(i) * (2 << depth)
  }
}

/// Returns the left and right most nodes in the tree that the node spans.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::spans(0), (0, 0));
/// assert_eq!(flat_tree::spans(1), (0, 2));
/// assert_eq!(flat_tree::spans(3), (0, 6));
/// assert_eq!(flat_tree::spans(23), (16, 30));
/// assert_eq!(flat_tree::spans(27), (24, 30));
/// ```
#[inline]
pub fn spans(i: u64) -> (u64, u64) {
  (left_span(i), right_span(i))
}

/// Returns how many nodes are in the tree that the node spans.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::count(0), 1);
/// assert_eq!(flat_tree::count(1), 3);
/// assert_eq!(flat_tree::count(3), 7);
/// assert_eq!(flat_tree::count(5), 3);
/// assert_eq!(flat_tree::count(23), 15);
/// assert_eq!(flat_tree::count(27), 7);
/// ```
#[inline]
pub const fn count(i: u64) -> u64 {
  let depth = self::depth(i);
  (2 << depth) - 1
}

/// Returns a list of all the full roots (subtrees where all nodes have either 2
/// or 0 children) `<` index.
///
/// For example `fullRoots(8)` returns `[3]` since the subtree rooted at `3`
/// spans `0 -> 6`, and the tree rooted at `7` has a child located at `9` which
/// is `>= 8`.
///
/// ## Panics
/// If an uneven index is passed.
///
/// ## Examples
/// ```rust
/// use flat_tree::full_roots;
///
/// let mut nodes = Vec::with_capacity(16);
/// full_roots(0, &mut nodes);
/// assert_eq!(nodes, []);
///
/// let mut nodes = Vec::with_capacity(16);
/// full_roots(2, &mut nodes);
/// assert_eq!(nodes, [0]);
///
/// let mut nodes = Vec::with_capacity(16);
/// full_roots(8, &mut nodes);
/// assert_eq!(nodes, [3]);
///
/// let mut nodes = Vec::with_capacity(16);
/// full_roots(20, &mut nodes);
/// assert_eq!(nodes, [7, 17]);
///
/// let mut nodes = Vec::with_capacity(16);
/// full_roots(18, &mut nodes);
/// assert_eq!(nodes, [7, 16]);
///
/// let mut nodes = Vec::with_capacity(16);
/// full_roots(16, &mut nodes);
/// assert_eq!(nodes, [7]);
/// ```
#[inline]
pub fn full_roots(i: u64, nodes: &mut Vec<u64>) {
  assert!(
    is_even(i),
    format!(
      "You can only look up roots for depth 0 blocks, got index {}",
      i
    )
  );
  let mut tmp = i >> 1;
  let mut offset = 0;
  let mut factor = 1;

  loop {
    if tmp == 0 {
      break;
    }
    while factor * 2 <= tmp {
      factor *= 2;
    }
    nodes.push(offset + factor - 1);
    offset += 2 * factor;
    tmp -= factor;
    factor = 1;
  }
}

#[inline]
pub(crate) const fn is_even(num: u64) -> bool {
  (num & 1) == 0
}

#[inline]
pub(crate) const fn is_odd(num: u64) -> bool {
  (num & 1) != 0
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn test_is_even() {
    assert_eq!(is_even(0), true);
    assert_eq!(is_even(1), false);
    assert_eq!(is_even(2), true);
    assert_eq!(is_even(3), false);
  }

  #[test]
  fn test_is_odd() {
    assert_eq!(is_odd(0), false);
    assert_eq!(is_odd(1), true);
    assert_eq!(is_odd(2), false);
    assert_eq!(is_odd(3), true);
  }

  #[test]
  fn test_parent_gt_int32() {
    assert_eq!(parent(10_000_000_000), 10_000_000_001);
  }

  #[test]
  fn test_child_to_parent_to_child() {
    let mut child = 0;
    for _ in 0..50 {
      child = parent(child)
    }
    assert_eq!(child, 1_125_899_906_842_623);
    for _ in 0..50 {
      child = left_child(child).expect("no valid number returned");
    }
    assert_eq!(child, 0);
  }
}
